core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala

```diff
@@ -79,12 +79,23 @@ private[spark] class CoarseGrainedExecutorBackend(
    */
   private[executor] val taskResources = new mutable.HashMap[Long, Map[String, ResourceInformation]]
 
-  @volatile private var decommissioned = false
+  private var decommissioned = false
 
   override def onStart(): Unit = {
-    logInfo("Registering PWR handler.")
-    SignalUtils.register("PWR", "Failed to register SIGPWR handler - " +
-      "disabling decommission feature.")(decommissionSelf)
+    if (env.conf.get(DECOMMISSION_ENABLED)) {
+      logInfo("Registering PWR handler to trigger decommissioning.")
+      SignalUtils.register("PWR", "Failed to register SIGPWR handler - " +
+      "disabling executor decommission feature.") {
+        self.send(DecommissionExecutor)
```

Comment:
@cloud-fan @agrawaldevesh What do you think about the latest change? In case of executor is decommissioned by the signal, we send `DecommissionExecutor`(it's a local RPC call) to the executor itself instead of invoking `decommissionSelf` directly.

In this way, we can guarantee the thread-safe of `decommissioned`.

(It's same to the Worker.)

Reply:
SGTM

Reply:
This is async, but I think it's safe here as the signal handler just needs to start decommissioning.

Reply:
I think being async here is fine, but I would like to better understand the need for inlining the logic (removed from L273 below in here).

Is ExecutorDecommissioning only supposed to be called when the executor receives a SIGPWR and NOT when it it is asked to decom by the driver (or in the standalone scenario) ? 

Earlier it seemed that both the PWR signal handling and the other codepaths went through the same `decommissionSelf` code flow but now it they appear a bit different.

Reply:
Okay I think I understand why ExecutorDecommissioning is only sent in this SIGPWR case: What do you think of introducing a new message like ExecutorSigPwrReceived and handle that in the event loop ? I am wary of doing all this work in the signal handling context. I would prefer it being done in the regular event loop.

Reply:
I'm ok with adding a new mesage.

WDYT? @cloud-fan

Reply:
good idea! This is clearer and also makes the signal handler logic simpler.